torch/csrc/jit/passes/constant_propagation.cpp

```diff
@@ -63,28 +61,124 @@ void propagateNode(Node* n) {
   }
 }
 
+void lowerIf(Block *body, Node * n) {
+  auto graph = n->owningGraph();
+  WithInsertPoint insert_point_guard { n };
+
+  std::unordered_map<Value*, Value*> value_map;
+  auto get_value = [&](Value *v) {
+    auto it = value_map.find(v);
+    if (it != value_map.end())
+      return it->second;
+    return v;
+  };
+
+  for (Node *orig : body->nodes()) {
+    Node *clone = graph->insertNode(graph->createClone(orig, get_value));
+    for (size_t i = 0; i < orig->outputs().size(); ++i) {
+      value_map[orig->outputs()[i]] = clone->outputs()[i];
+    }
+  }
+  for (size_t i = 0; i < n->outputs().size(); ++i) {
+    n->outputs().at(i)->replaceAllUsesWith(get_value(body->outputs().at(i)));
+  }
+  // NB: destroy the node here, because it might contain side effects, like print
+  n->destroy();
+}
+
+bool isTrueConstant(Value *val) {
+  at::optional<bool> maybe_value = constant_as<bool>(val);
+  return maybe_value && *maybe_value;
+}
+
+void lowerIf(Node *n) {
+  if (isTrueConstant(n->input())) {
+    lowerIf(n->blocks()[0], n);
+  } else {
+    lowerIf(n->blocks()[1], n);
+  }
+}
+
+//returns true if the mutated variables are changed
+bool recomputeMutatedVariables(Node *n) {
```

Comment:
I don't understand what is going on here. All variables are in SSA form, so no variable is every mutated. I think this is doing something else, but I could use an explanation.

Reply:
Sorry yea it's recomputing which values should be returned by the if node. If a node used to return (a, b) calculate if it should now only return (a), and return true if the values changed.

I was following the convention of compiler.cpp in emitIf which calls them mutated_variables.